pyanaconda/modules/subscription/installation.py

```diff
@@ -104,3 +108,150 @@ def run(self):
+        log in debug mode also on the system once rhsm.conf has been
+        copied over to the target system.
+
+        So set the log level back to INFO before we copy the config file.
+        """
+        log.debug("subscription: setting RHSM log level back to INFO")
+        self._rhsm_config_proxy.Set("logging.default_log_level",
+                                    get_variant(Str, "INFO"), "")
+
+
+class TransferSubscriptionTokensTask(Task):
+    """Transfer subscription tokens to the target system."""
+
+    RHSM_REPO_FILE_PATH = "/etc/yum.repos.d/redhat.repo"
+    RHSM_CONFIG_FILE_PATH = "/etc/rhsm/rhsm.conf"
+    RHSM_SYSPURPOSE_FILE_PATH = "/etc/rhsm/syspurpose/syspurpose.json"
+    RHSM_ENTITLEMENT_KEYS_PATH = "/etc/pki/entitlement"
+    RHSM_CONSUMER_KEY_PATH = "/etc/pki/consumer/key.pem"
+    RHSM_CONSUMER_CERT_PATH = "/etc/pki/consumer/cert.pem"
+
+    TARGET_REPO_FOLDER_PATH = "/etc/yum.repos.d"
+
+    def __init__(self, sysroot, transfer_subscription_tokens):
+        """Create a new task.
+
+        :param str sysroot: target system root path
+        :param bool transfer_subscription_tokens: if True attempt to transfer subscription
+                                                  tokens to target system (we always transfer
+                                                  system purpose data unconditionally)
+        """
+        super().__init__()
+        self._sysroot = sysroot
+        self._transfer_subscription_tokens = transfer_subscription_tokens
+
+    @property
+    def name(self):
+        return "Transfer subscription tokens to target system"
+
+    def _copy_pem_files(self, input_folder, output_folder, not_empty=True):
+        """Copy all pem files from input_folder to output_folder.
+        Files with the pem extension are generally encryption keys and certificates.
+        If output_folder does not exist, it & any parts of its path will
+        be created.
+        :param str input_folder: input folder for the pem files
+        :param str output_folder: output folder where to copy the pem files
+        :return: False if the input directory does not exists or is empty,
+                True after all pem files have be successfully copied
+        :rtype: bool
+        """
+        # check the input folder exists
+        if not os.path.isdir(input_folder):
+            return False
+        # optionally check the input folder is not empty
+        if not_empty and not os.listdir(input_folder):
+            return False
+        # make sure the output folder exist
+        util.mkdirChain(output_folder)
+        # transfer all the pem files in the input folder
+        for pem_file_path in glob.glob(os.path.join(input_folder, "*.pem")):
+            shutil.copy(pem_file_path, output_folder)
+        # if we got this far the pem copy operation was a success
+        return True
+
+    def _copy_file(self, file_path, target_file_path):
```

Comment:
What the task does is copying a bunch of files from the installation environment to the target system in similar yet slightly different ways.

So there could either be a whole bunch of pretty similar pieces of code, doing the slightly different things for each file/bunch of files, which would likely not look very good and would be cumbersome to unit test.

So instead an effort has been made to extract the similar pieces to private helper methods to reduce code duplication & make the task easier to test.

For the individual methods:

`_copy_pem_files()` - this copies `*.pem` files related to the attached subscription and needs to handle extension based filtering (so that only `pem` files are transferred), the input folder not existing as well as making sure the output folder exists

`_copy_file()` - general file copy function that checks if the target is file and makes sure the target folder exists

`_transfer_file()` - nice wrapper for `_copy_file()` with logging and exception raising based on what is being transferred

`_ transfer_system_purpose()` - wrapper for `_copy_file()` with system purpose specifics, mainly due to missing system purpose data not being a fatal error unlike ale the other files

`_transfer_entitlement_keys()` - wrapper for `_copy_pem_files()` with logging and exception raising

You could unify some of these methods (eq. copy-past _copy_file() in place to `_transfer_file()` and `_transfer_system purpose()` and roll in the pem file transfer method to `_transfer_entitlement_keys()`) but that would IMHO be at the cost of complexity and harder unit testing.